parser/src/stateNode.ts

```diff
@@ -0,0 +1,114 @@
+import * as t from "@babel/types";
+import { MaybeArrayOfActions } from "./actions";
+import { Context } from "./context";
+import { History } from "./history";
+import { Invoke } from "./invoke";
+import { StateMeta } from "./meta";
+import {
+  AnyNode,
+  BooleanLiteral,
+  NumericLiteral,
+  StringLiteral,
+} from "./scalars";
+import { Schema } from "./schema";
+import { MaybeTransitionArray } from "./transitions";
+import { TsTypes } from "./tsTypes";
+import { AnyParser } from "./types";
+import {
+  GetObjectKeysResult,
+  GetParserResult,
+  maybeArrayOf,
+  objectOf,
+  ObjectOfReturn,
+  ObjectPropertyInfo,
+  objectTypeWithKnownKeys,
+} from "./utils";
+
+const On = objectOf(MaybeTransitionArray);
+
+const After = objectOf(MaybeTransitionArray);
+const Tags = maybeArrayOf(StringLiteral);
+
+/**
+ * This is frustrating, but we need to keep this
+ * up to date with the StateNode definition below.
+ *
+ * The reason? TS fails early when it hits a
+ * recursive type definition, meaning our inference
+ * falls out the window when StateNode tries to
+ * reference itself
+ */
+export type StateNodeReturn = {
+  id?: GetParserResult<typeof StringLiteral>;
+  initial?: GetParserResult<typeof StringLiteral>;
+  type?: GetParserResult<typeof StringLiteral>;
+  delimiter?: GetParserResult<typeof StringLiteral>;
+  entry?: GetParserResult<typeof MaybeArrayOfActions>;
+  exit?: GetParserResult<typeof MaybeArrayOfActions>;
+  onEntry?: GetParserResult<typeof MaybeArrayOfActions>;
+  onExit?: GetParserResult<typeof MaybeArrayOfActions>;
+  invoke?: GetParserResult<typeof Invoke>;
+  always?: GetParserResult<typeof MaybeTransitionArray>;
+  onDone?: GetParserResult<typeof MaybeTransitionArray>;
+  on?: GetParserResult<typeof On>;
+  after?: GetParserResult<typeof After>;
+  history?: GetParserResult<typeof History>;
+  tags?: GetParserResult<typeof Tags>;
+  states?: GetParserResult<AnyParser<ObjectOfReturn<StateNodeReturn>>>;
+  node: t.Node;
+  meta?: GetParserResult<typeof StateMeta>;
+  tsTypes?: GetParserResult<typeof TsTypes>;
+  context?: GetParserResult<typeof Context>;
+  data?: GetParserResult<typeof AnyNode>;
+  schema?: GetParserResult<typeof Schema>;
+  preserveActionOrder?: GetParserResult<typeof BooleanLiteral>;
+  parallel?: GetParserResult<typeof BooleanLiteral>;
```

Comment:
This got superseded by `type: 'parallel'` quite a long time ago so we probably don't need to support this
```suggestion
```

Reply:
We HAVE to support everything that's in the TS declaration, otherwise TS yells at us.